src/main/java/com/googleintern/wfm/ruleengine/action/KarnaughMapTermGenerator.java

```diff
@@ -0,0 +1,148 @@
+      index = index + 1;
+    }
+    return filterByIndexBuilder.build();
+  }
+
+  /**
+   * Find all combinations of filters that are not covered by the input data set.
+   *
+   * <p>Steps:
+   *
+   * <ol>
+   *   <li>Step 1: Find all possible combinations of terms using Filter index in BiMap. Save
+   *       generated results in allFilterPermutations.
+   *   <li>Step 2: Find all one terms. One terms are defined as existing terms(Filter combinations)
+   *       and covered terms from input data set. Save finding results in allOneTerms.
+   *   <li>Step 3: Find all zero terms. Zero terms are terms that are not covered by the input data
+   *       set.
+   * </ol>
+   */
+  public static ImmutableSet<ImmutableList<Integer>> findAllZeroTerms(
+      ImmutableBiMap<FilterModel, Integer> filterByIndex,
+      ImmutableSet<ImmutableList<FilterModel>> filters) {
+    int totalNumberOfFilters = filterByIndex.keySet().size();
+    List<Integer> filterNumberAllZeros =
+        new ArrayList<>(Collections.nCopies(totalNumberOfFilters, 0));
+    ImmutableSet<ImmutableList<Integer>> allFilterPermutations =
+        findAllPossibleTerms(filterNumberAllZeros, totalNumberOfFilters - 1);
+    ImmutableSet<ImmutableList<Integer>> allOneTerms =
+        findAllOneTerms(filterByIndex, allFilterPermutations, filters);
+    return ImmutableSet.copyOf(Sets.difference(allFilterPermutations, allOneTerms));
+  }
+
+  private static ImmutableSet<FilterModel> findAllTypesOfFilters(
+      ImmutableSet<ImmutableList<FilterModel>> filters) {
+    if (filters.size() == 0) {
+      return ImmutableSet.of();
+    }
+    ImmutableSet.Builder<FilterModel> allTypesOfFiltersBuilder = ImmutableSet.builder();
+    filters.forEach(
+        filterList -> filterList.forEach(filter -> allTypesOfFiltersBuilder.add(filter)));
+    return allTypesOfFiltersBuilder.build();
+  }
+
+  private static ImmutableSet<ImmutableList<Integer>> findAllIncludedTerms(
+      ImmutableBiMap<FilterModel, Integer> filterByIndex,
+      ImmutableSet<ImmutableList<Integer>> allFilterPermutations,
+      ImmutableList<FilterModel> filters) {
+    ImmutableSet.Builder<Integer> onesIndexBuilder = ImmutableSet.builder();
+    for (FilterModel filter : filters) {
+      onesIndexBuilder.add(filterByIndex.get(filter));
+    }
+    ImmutableSet<Integer> onesIndex = onesIndexBuilder.build();
+    ImmutableSet.Builder<ImmutableList<Integer>> allIncludedCasesBuilder = ImmutableSet.builder();
+    for (ImmutableList filterPermutation : allFilterPermutations) {
+      if (isTermIncluded(filterPermutation, onesIndex)) {
+        allIncludedCasesBuilder.add(filterPermutation);
+      }
+    }
+    return allIncludedCasesBuilder.build();
+  }
+
+  private static boolean isTermIncluded(
+      ImmutableList<Integer> filterPermutation, ImmutableSet<Integer> onesIndex) {
+    for (Integer index : onesIndex) {
+      if (filterPermutation.get(index) == 0) {
+        return false;
+      }
+    }
+    return true;
```

Comment:
this can be done with streams
```suggestion
    return onesIndex.stream().noneMatch(index -> filterPermutation.get(index) == 0);
```